doc/tutorial/extending_theano.txt

```diff
@@ -286,6 +297,36 @@ Op Example
                 return eval_points
             return self.grad(inputs, eval_points)
 
+.. testcode:: example (using itypes and otypes)
+
+    import theano
+
+    class DoubleOp(theano.Op):
+        __props__ = ()
+
+        itypes = [theano.tensor.dmatrix]
+        otypes = [theano.tensor.dmatrix]
+
+        def perform(self, node, inputs, output_storage):
+            x = inputs[0]
+            z = output_storage[0]
+            z[0] = x * 2
+
+        def infer_shape(self, node, i0_shapes):
+            return i0_shapes
+
+        def grad(self, inputs, output_grads):
+            return [output_grads[0] * 2]
+
+        def R_op(self, inputs, eval_points):
+            # R_op can receive None as eval_points.
+            # That mean there is no diferientiable path through that input
+            # If this imply that you cannot compute some outputs,
+            # return None for those.
+            if eval_points[0] is None:
+                return eval_points
+            return self.grad(inputs, eval_points)
+
```

Comment:
Yes, both testcode section need to get tested. So mostly,

testcode define op version one
testcode example
testoutput
testcode define op version 2
testcode example that use it
testoutput

Or it can be that you define the 2 class with different class name and test them together? This would show more clearly that it is equivalent I think.